### PR TITLE
PIM-5701 && PIM-5705: add CSV and XLSX imports and exports for channels, locales, currencies, group types and attribute groups

### PR DESCRIPTION
**Pull request by aRn0D**
| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | yes
| Added Behats                      | yes
| Changelog updated                 | yes
| Review and 2 GTM                  | yes
| Micro Demo to the PO (Story only) | TODO
| Migration script                  |  no
| Tech Doc                          | no

### CODE REVIEW

File: features/Context/FixturesContext.php [hunk 1]
```diff
@@ -414,6 +415,100 @@ public function thereShouldBeTheFollowingFamilies(TableNode $table)
...
+     * @param TableNode $table
+     *
+     * @Then /^there should be the following group types:$/
+     */
+    public function thereShouldBeTheFollowingGroupTypes(TableNode $table)
+    {
+        foreach ($table->getHash() as $data) {
+            $groupType = $this->getGroupType($data['code']);
+
+            assertEquals($data['label-en_US'], $groupType->getTranslation('en_US')->getLabel());
```
Comment (grena) [inline features/Context/FixturesContext.php]: I don't understand why this assertion here?

File: features/Context/FixturesContext.php [hunk 2]
```diff
@@ -414,6 +415,100 @@ public function thereShouldBeTheFollowingFamilies(TableNode $table)
...
+     *
+     * @Then /^there should be the following attribute groups:$/
+     */
+    public function thereShouldBeTheFollowingAttributeGroups(TableNode $table)
+    {
+        foreach ($table->getHash() as $data) {
+            /** @var AttributeGroup $group */
+            $group = $this->getAttributeGroup($data['code']);
+
+            assertEquals($data['label-en_US'], $group->getTranslation('en_US')->getLabel());
```
Comment (grena) [inline features/Context/FixturesContext.php]: same question, why this assertion?

File: features/export/export_locales.feature [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    vi_VN
+    wo_SN
+    xh_ZA
+    yo_NG
+    zh_CN
+    zh_HK
+    zh_MO
+    zh_SG
+    zh_TW
+    zu_ZA
```
Comment (grena) [inline features/export/export_locales.feature]: What a long list :(

Comment (aRn0D) [inline features/export/export_locales.feature]: This is the default locale :)

File: src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php [hunk 4]
```diff
@@ -20,12 +21,21 @@ class AttributeGroupUpdater implements ObjectUpdaterInterface
     /** @var IdentifiableObjectRepositoryInterface */
     protected $attributeRepository;
 
+    /** @var AttributeGroupManager */
+    protected $attributeGroupManager;
+
     /**
+     * AttributeGroupUpdater constructor.
```
Comment (grena) [inline src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php]: please remove

File: src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php [hunk 5]
```diff
@@ -74,6 +84,11 @@ protected function setData($attributeGroup, $field, $data)
         } elseif ('sort_order' == $field) {
             $attributeGroup->setSortOrder($data);
         } elseif ('attributes' == $field) {
+            $attributes = $attributeGroup->getAttributes();
+            foreach ($attributes as $attribute) {
+                $this->attributeGroupManager->removeAttribute($attributeGroup, $attribute);
```
Comment (grena) [inline src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php]: can't we use an AttributeRemover instead of the manager?

**Comment by grena** [comment]
- **The migration script to add new jobs is missing.**
- Question about the `AttributeManager`, to see if we can use the Remover instead.
- I don't like the giant behats (more than 200 lines of CSV), but I think we don't have the choice

Then GTM for me.

File: features/Context/FixturesContext.php (hunk 1 above)
Comment (aRn0D) [inline features/Context/FixturesContext.php]: Because of : 
```
Then there should be the following group types:
      | code         | label-en_US | is_variant |
      | cross_sell | Cross sell     | 0              |
```

File: features/Context/FixturesContext.php (hunk 2 above)
Comment (aRn0D) [inline features/Context/FixturesContext.php]: Because of 
```
Then there should be the following attribute groups:
  | code          | label-en_US   | attributes       |
  | manufacturing | Manufacturing | lace_fabric |
```

**Comment by aRn0D** [comment]
We don't need migration :)

File: src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php (hunk 5 above)
Comment (aRn0D) [inline src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php]: Because the `AttributeGroupRemover` does not exist. you should remove this manager but it should be done in another card. We need to delete this manager.

File: src/Pim/Bundle/ConnectorBundle/Resources/translations/messages.en.yml [hunk 6]
```diff
@@ -253,3 +288,38 @@ batch_jobs:
     xlsx_association_type_export:
         label: Association type export in XLSX
         export.label: Association type export in XLSX
+    xlsx_channel_export:
+        label: Channel export in CSV
```
Comment (damien-carcel) [inline src/Pim/Bundle/ConnectorBundle/Resources/translations/messages.en.yml]: In XLSX
Same problem for all the followings

**Comment by damien-carcel** [comment]
Small comments then GTM

File: features/Context/FixturesContext.php [hunk 7]
```diff
@@ -414,6 +415,101 @@ public function thereShouldBeTheFollowingFamilies(TableNode $table)
     }
 
     /**
+     * @param TableNode $table
+     *
+     * @Then /^there should be the following currencies:$/
```
Comment (SamirBoulil) [inline features/Context/FixturesContext.php]: "There should be the following *activated* currencies"  is what you test, isn't it ?

File: features/Context/FixturesContext.php [hunk 8]
```diff
@@ -414,6 +415,101 @@ public function thereShouldBeTheFollowingFamilies(TableNode $table)
...
+     *
+     * @Then /^there should be the following channels:$/
+     */
+    public function thereShouldBeTheFollowingChannels(TableNode $table)
+    {
+        foreach ($table->getHash() as $data) {
+            $channel = $this->getChannel($data['code']);
+
+            assertEquals($data['label'], $channel->getLabel());
+            assertEquals($data['color'], $channel->getColor());
```
Comment (SamirBoulil) [inline features/Context/FixturesContext.php]: I thought channels didn't have a color anymore ?

File: features/Context/catalog/footwear/jobs.yml [hunk 9]
```diff
@@ -160,6 +160,56 @@ jobs:
...
+    csv_footwear_currency_export:
+        connector: Akeneo CSV Connector
+        alias:     csv_currency_export
+        label:     CSV footwear currency export
+        type:      export
+        configuration:
+            delimiter:  ;
+            enclosure:  '"'
+            withHeader: true
+            filePath: /tmp/currency.csv
```
Comment (SamirBoulil) [inline features/Context/catalog/footwear/jobs.yml]: align ;)

File: features/Context/catalog/footwear/jobs.yml [hunk 10]
```diff
@@ -160,6 +160,56 @@ jobs:
...
+    csv_footwear_channel_export:
+        connector: Akeneo CSV Connector
+        alias:     csv_channel_export
+        label:     CSV footwear channel export
+        type:      export
+        configuration:
+            delimiter:  ;
+            enclosure:  '"'
+            withHeader: true
+            filePath: /tmp/channel.csv
```
Comment (SamirBoulil) [inline features/Context/catalog/footwear/jobs.yml]: align ;)

File: src/Pim/Bundle/CatalogBundle/Entity/AttributeGroup.php [hunk 11]
```diff
@@ -175,10 +175,8 @@ public function setUpdated($updated)
      */
     public function addAttribute(AttributeInterface $attribute)
     {
-        if (!$this->hasAttribute($attribute)) {
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Entity/AttributeGroup.php]: Why did you remove the check here ?

File: src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php [hunk 12]
```diff
@@ -74,16 +82,19 @@ protected function setData($attributeGroup, $field, $data)
         } elseif ('sort_order' == $field) {
             $attributeGroup->setSortOrder($data);
         } elseif ('attributes' == $field) {
+            foreach ($attributeGroup->getAttributes() as $attribute) {
+                $this->attributeGroupManager->removeAttribute($attributeGroup, $attribute);
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php]: oulah
nope nope nope

you're in the Updater, which means you only modify your object in memory
in the manager you save it to the database

you shouldn't do that imho

File: features/Context/FixturesContext.php (hunk 1 above)
Comment (skeleton) [inline features/Context/FixturesContext.php]: what's append if there is `label-fr_FR` ?

File: features/import/xlsx/import_channels.feature [hunk 13]
```diff
@@ -18,5 +18,25 @@ Feature: Import channels
...
+      | code | label | color | currencies | locales           | tree            | conversion_units |
+      | site | Site  | blue  | EUR,USD    | de_DE,en_US,hy_AM | 2014_collection |                  |
+
+  Scenario: Successfully remove locales and currencies
+    Given the "apparel" catalog configuration
+    And the following jobs:
+      | connector             | type   | alias               | code                | label               |
+      | Akeneo XLSX Connector | import | xlsx_channel_import | xlsx_channel_import | XLSX channel import |
+    And I am logged in as "Julia"
+    And the following XLSX file to import:
```
Comment (SamirBoulil) [inline features/import/xlsx/import_channels.feature]: Does an XLSX file specified this way works for XLSX jobs ? We usually use the table notation for xlsx imports.

File: src/Pim/Component/Connector/Writer/Database/AttributeGroupWriter.php [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function write(array $objects)
+    {
+        $this->incrementCount($objects);
+
+        if (null !== $defaultGroup = $this->attributeGroupRepository->findDefaultAttributeGroup()) {
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/Database/AttributeGroupWriter.php]: please explain what you do and why you do it in the phpdoc of the class

File: src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php [hunk 15]
```diff
@@ -113,4 +101,34 @@ protected function findAttribute($attributeCode)
     {
         return $this->attributeRepository->findOneByIdentifier($attributeCode);
     }
+
+    /**
+     * @param AttributeGroupInterface $attributeGroup
+     * @param string[]                $data
+     */
+    protected function setAttribute(AttributeGroupInterface $attributeGroup, array $data)
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php]: I don't understand the naming of the method

File: src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php [hunk 16]
```diff
@@ -113,4 +101,34 @@ protected function findAttribute($attributeCode)
...
+    {
+        if ('other' === $attributeGroup->getCode()) {
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php]: don't we have a constant for that?

File: src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php (hunk 15 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php]: setAttributes

File: features/Context/FixturesContext.php (hunk 1 above)
Comment (aRn0D) [inline features/Context/FixturesContext.php]: We need to add a new check

File: src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php (hunk 16 above)
Comment (aRn0D) [inline src/Pim/Component/Catalog/Updater/AttributeGroupUpdater.php]: no :(
